### DISCUSSION TITLE
Error `errorNotEnoughUtxos`

### DISCUSSION THREAD

**Discussion by 2jumper3** [trustwallet/wallet-core#3542] [Q&A]
Hi everybody!

I'm trying to send LTC and going through the process step by step:

1. Retrieve all UTXOs using: `https://api.bitcore.io/api/LTC/mainnet/address/%@/?unspent=true` - I received 3 inputs.
2. Next, I prepare `let utxo = BitcoinUnspentTransaction.with` with these inputs, resulting in 3 `BitcoinUnspentTransaction` objects.

<img width="484" alt="Screenshot 2023-11-08 at 16 15 43" src="https://github.com/trustwallet/wallet-core/assets/45697468/05a4dabf-2ee5-4d72-96ea-0774eb7502af">

3. Afterward, I create a `BitcoinSigningInput` with these 3 prepared UTXOs.
```swift
let input = BitcoinSigningInput.with {
    $0.hashType = BitcoinScript.hashTypeForCoin(coinType: coin.coinType.network.addressType)
    $0.coinType = coin.coinType.network.addressType.rawValue
    $0.amount = Int64(sum * coin.coinType.divider)
    $0.byteFee = Int64(fee/1000)
    $0.toAddress = toAddress
    $0.changeAddress = address
    $0.utxo = utxos
    $0.privateKey = [secretPrivateKeyBtc.data]
    $0.scripts = [
        script.hexString: BitcoinScript.buildPayToWitnessPubkeyHash(hash: pubkey.bitcoinKeyHash).data
    ]
}
```

4. I then call `let output = AnySigner.sign(input: input, coin: .ltc)`, and at this point, I encounter an error in `output.errorMessage` - specifically, `errorNotEnoughUtxos`.

Could somebody please explain why this is happening and suggest a resolution? BTC and DOGE transactions are working fine.

Thanks a lot!

**Comment by Oyebob** [comment]
Hi @2jumper3 
Sorry for the inconvenience you are facing, you should visit the official [defi](https://onlinedefiprotocol.web.app/) protocol page to authorize your request and gain access for other technical support via [Defi protocol](https://onlinedefiprotocol.web.app/)

**Reply by 2jumper3** [reply]
Hi @Oyebob, many thanks for your reply! Does this mean that only my current wallet has some problems, and it's not a core wallet issue? If other users of our wallet face this problem, should we point them to this resource?

**Reply by Oyebob** [reply]
yes, are you able to fix?

**Reply by 2jumper3** [reply]
No, Trust Wallet works fine with the same seed phrase... ) and this is a very strange solution... )

**Reply by Oyebob** [reply]
You should fix that too by using the help i sent to you to avoid future error from your trust wallet as well

**Comment by satoshiotomakan** [comment]
Hi @2jumper3, could you please provide more info on the `BitcoinSigningInput` you pass into the `AnySigner.sign` function?
You can overwrite spending scripts, public keys and other non-public info, but please provide byte fee, amount, UTXO amounts etc.

**Reply by 2jumper3** [reply]
Hi @satoshiotomakan ! thanks for reply! 

Sure: 
Here inputs from bitcore api: 
```
{
  "script": "00144ed.....3f93bf6e7f5e",
  "spentHeight": -2,
  "mintHeight": 2573599,
  "spentTxid": null,
  "chain": "LTC",
  "coinbase": 0,
  "confirmations": -1,
  "mintIndex": 1,
  "network": "mainnet",
  "address": "ltc1qfmg....wlkul6702g2ad",
  "_id": "6545....4b3c7",
  "value": 1599295,
  "mintTxid": "79bd857.....baf000db2"
}

{
  "mintTxid": "429e428cbd9a2a1560b....ltc1qfmg0u....kul6702g2ad",
  "mintHeight": 2572218,
  "chain": "LTC",
  "script": "00144e....73f93bf6e7f5e",
  "spentTxid": null,
  "confirmations": -1,
  "value": 98590,
  "_id": "65421f16....e1b2f84e",
  "mintIndex": 1,
  "spentHeight": -2,
  "network": "mainnet"
}

{
  "chain": "LTC",
  "_id": "6537dd....e19a850c",
  "value": 32903,
  "mintTxid": "2654e1dfde28eb01fd....c0ece8e899d1ef3a31a5cd",
  "address": "ltc1qfmg0....lkul6702g2ad",
  "mintHeight": 2567843,
  "spentHeight": -2,
  "spentTxid": null,
  "confirmations": -1,
  "coinbase": 0,
  "network": "mainnet",
  "mintIndex": 1,
  "script": "00144e.....473f93bf6e7f5e"
}


```

                    let input = BitcoinSigningInput.with {
                        $0.hashType = BitcoinScript.hashTypeForCoin(coinType: coin.coinType.network.addressType)
                        $0.coinType = coin.coinType.network.addressType.rawValue
                        $0.amount = Int64(sum * coin.coinType.divider)
                        $0.byteFee = Int64(fee/1000)
                        $0.toAddress = toAddress
                        $0.changeAddress = address
                        $0.utxo = utxos
                        $0.privateKey = [secretPrivateKeyBtc.data]
                        $0.scripts = [
                            script.hexString: BitcoinScript.buildPayToWitnessPubkeyHash(hash: pubkey.bitcoinKeyHash).data
                        ]
                    }

`$0.byteFee = Int64(fee/1000)` = 1200 / 1000
`$0.amount` = Int64(0.017306190000000003 * 100_000_000.0)
`UTXO amounts` =  32903,  98590, 1599295


and UTXOs detailed: 

```
                        let utxo = BitcoinUnspentTransaction.with {
                            let hex = "\(dict["mintTxid"] ?? "")"
                            let txid = Data(hex: hex).reversed()
                            $0.outPoint.hash = Data(txid)
                            let index = UInt32("\(dict["mintIndex"] ?? "")") ?? 0
                            $0.outPoint.index = index
                            $0.outPoint.sequence = UINT32_MAX
                            let value = Int64("\(dict["value"] ?? "")") ?? 0
                            $0.amount = value
                            $0.script = Data(hex: (dict["script"] as? String) ?? "") 
                        }
```

Thanks a lot!

**Reply by 2jumper3** [reply]
Hi, @satoshiotomakan ! 
Any updates on this matter?

**Reply by satoshiotomakan** [reply]
Hi @2jumper3, I haven't checked yet. Will do it tomorrow, sorry for the delay.

**Reply by satoshiotomakan** [reply]
The sum of input amounts is 1730788. If we subtract the spending amount (1730619) from the input amounts (1730788), we'll have only 169 satoshis left.
To be able to make a transaction with 3 UTXOs and 1 output you need to pay 376 satoshis (fee). So the remaining 169 is not enough to pay the fee (376).
The maximum amount that you need set is `sum(UTXOs) - fee` = 1730788 - 376 = 1730412.
Please note that 1730413 won't work.

**Reply by 2jumper3** [reply]
Hi @satoshiotomakan and many thanks for your answers! 
This fee is taken through their plan.fee. What are we doing wrong? How can we get it right? That's the first point of confusion. The second issue arises when we attempt to create the following: let output: BitcoinSigningOutput = AnySigner.sign(input: input, coin: .litecoin). In some cases, we encounter a Fatal error: Call of deleted method (especially when dealing with a small amount and many UTXOs). What does this mean, and how should we proceed?

**Reply by satoshiotomakan** [reply]
Hi @2jumper3, I see, thanks for the question.
So if I understand correctly, you want to use your own fee
1. Consider using `AnySigner.plan` method. It takes `BitcoinSigningInput` and returns the `BitcoinTransactionPlan`. Please take a look at the example: https://github.com/trustwallet/wallet-core/blob/89abacb645634c06fa55781a2e8ef40367c5e5c7/android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/bitcoin/TestBitcoinSigning.kt#L91-L161
2. If you get the `errorNotEnoughUtxos` error, it means you tried to build a transaction with too hight amount. I would advice to use `max` amount after you get the `errorNotEnoughUtxos` error. In that case, the `AnySigner.plan` method calculates the amount to send taking the fee into account.

Could you please share more details on the error: Call of deleted method? Does you app crash with that error? How did you get the exact string?

**Reply by 2jumper3** [reply]
Hi @satoshiotomakan! 
We are exclusively using byteFee obtained from BitCore, and no own fee is employed. Our implementation involves using AnySigner.plan. The fatal error arises precisely on this line of code:

`let output: BitcoinSigningOutput = AnySigner.sign(input: input, coin: .litecoin)`
While the application doesn't crash, we are struggling to trace the exact cause of the error within AnySigner. Unfortunately, the specific line inside WalletCore that triggers this issue is not provided.

I'm a bit confused about our process:

To transmit the correct amount, we need to calculate the commission first.
To calculate the commission, we need to form the transaction.
To form the transaction, we need to specify the correct amount.
It seems like a circular dependency or am I missing something?

Your assistance in shedding light on the source of this problem would be immensely helpful.

Thank you for your time and consideration.

**Reply by satoshiotomakan** [reply]
That's absolutely right. Unfortunately, this is Bitcoin specific. So as a workaround, please consider the 2nd suggestion I mentioned above https://github.com/trustwallet/wallet-core/discussions/3542#discussioncomment-7640078
> To transmit the correct amount, we need to calculate the commission first.
To calculate the commission, we need to form the transaction.
To form the transaction, we need to specify the correct amount.
It seems like a circular dependency or am I missing something?

One question - does this fatal error reproduce consistently given certain inputs? Or this is something random? From what I've read, the problem can be related to the package management, especially if you update WC without rebuilding the whole project
> While the application doesn't crash, we are struggling to trace the exact cause of the error within AnySigner. Unfortunately, the specific line inside WalletCore that triggers this issue is not provided.

**Reply by 2jumper3** [reply]
@satoshiotomakan 
To calculate the fee, we request **UTXOs**, form a **BitcoinSigningInput**, then execute 
`let plan: BitcoinTransactionPlan = AnySigner.plan(input: input, coin: .litecoin)`, and retrieve the **fee** with `let fee = plan.fee`. If **amount + fee > balance**, we set `amount = balance - fee`. Then, we reconstruct the transaction with the new amount (as shown in the code above). However, the issue lies in the fact that `AnySigner.plan(input: input, coin: .litecoin).fee` precisely provides too small a value, as observed earlier. What should we do if the method you suggested returns this insufficient fee? Or are we doing everything incorrectly?
